From 2c6f115a508be7a8052baa1c2a39e3ff820a6bf7 Mon Sep 17 00:00:00 2001
From: Ramya Gnanasekar <quic_rgnanase@quicinc.com>
Date: Thu, 6 Apr 2023 22:52:34 +0530
Subject: [PATCH] mac80211: Fix 6 GHz Band capabilities IE advertisement

While adding 6GHz band capabilities, interface support and channel
is list is validated to confirm whether 6GHz is supported.
But in single wiphy, all the interfaces will have all the supported
channels. Due to this validation, mesh beacons adds
6 GHZ Band capabilities in beacon for 5GHz as well.

Fix it by adding 6 GHz band check before adding the IE.

Signed-off-by: Ramya Gnanasekar <quic_rgnanase@quicinc.com>

Index: b/net/mac80211/mesh.c
===================================================================
--- a/net/mac80211/mesh.c	2023-04-13 15:27:33.052959371 +0530
+++ b/net/mac80211/mesh.c	2023-04-13 15:35:09.385450821 +0530
@@ -645,7 +645,8 @@ int mesh_add_he_6ghz_cap_ie(struct ieee8
 	if (!iftd)
 		return 0;
 
-	ieee80211_ie_build_he_6ghz_cap(sdata, skb);
+	if (sband->band == NL80211_BAND_6GHZ)
+		ieee80211_ie_build_he_6ghz_cap(sdata, skb);
 	return 0;
 }
 
Index: b/net/mac80211/mlme.c
===================================================================
--- a/net/mac80211/mlme.c	2023-04-13 15:27:33.052959371 +0530
+++ b/net/mac80211/mlme.c	2023-04-13 15:28:46.452399833 +0530
@@ -721,7 +721,8 @@ static void ieee80211_add_he_ie(struct i
 	/* trim excess if any */
 	skb_trim(skb, skb->len - (pre_he_pos + he_cap_size - pos));
 
-	ieee80211_ie_build_he_6ghz_cap(sdata, skb);
+	if (sband->band == NL80211_BAND_6GHZ)
+		ieee80211_ie_build_he_6ghz_cap(sdata, skb);
 }
 
 static void ieee80211_add_eht_ie(struct ieee80211_sub_if_data *sdata,
